File: 500.600/547.NumberOfProvinces.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class NumberOfProvinces
{
public:
    int findCircleNum(const vector<vector<int>> &isConnected)
    {
        int n = isConnected.size();
        visited = vector<int>(n, 0);
        int ans = 0;
        for (int i = 0; i < n; ++i)
        {
            if (!visited[i])
            {
                dfs(isConnected, n, i);
                ++ans;
            }
        }
        return ans;
    }

    int findCircleNum1(const vector<vector<int>> &isConnected)
    {
        int n = isConnected.size();
        vector<int> visited(n, 0);
        int ans = 0;
        queue<int> qu;
        for (int i = 0; i < n; ++i)
        {
            if (!visited[i])
            {
                qu.push(i);
                while (!qu.empty())
                {
                    int j = qu.front();
                    qu.pop();
                    visited[j] = 1;
                    for (int k = 0; k < n; ++k)
                    {
                        if (isConnected[j][k] == 1 && !visited[k])
                        {
                            qu.push(k);
                        }
                    }
                }
                ++ans;
            }
        }
        return ans;
    }

    int findCircleNum2(const vector<vector<int>> &isConnected)
    {
        int n = isConnected.size();
        vector<int> parent(n);
        for (int i = 0; i < n; ++i)
        {
            parent[i] = i;
        }
        for (int i = 0; i < n; ++i)
        {
            for (int j = i + 1; j < n; ++j)
            {
                if (isConnected[i][j] == 1)
                    Union(parent, i, j);
            }
        }
        int ans = 0;
        for (int i = 0; i < n; ++i)
        {
            if (parent[i] == i)
                ++ans;
        }
        return ans;
    }

private:
    void dfs(const vector<vector<int>> &isConnected, int n, int i)
    {
        for (int j = 0; j < n; ++j)
        {
            if (isConnected[i][j] == 1 && !visited[j])
            {
                visited[j] = 1;
                dfs(isConnected, n, j);
            }
        }
    }

private:
    vector<int> visited;

private:
    int Find(vector<int> &parent, int index)
    {
        if (parent[index] != index)
        {
            parent[index] = Find(parent, parent[index]);
        }
        return parent[index];
    }
    void Union(vector<int> &parent, int index1, int index2)
    {
        parent[Find(parent, index1)] = Find(parent, index2);
    }
};

int main(int argc, char const *argv[])
{
    NumberOfProvinces cls;
    cout << cls.findCircleNum2(vector<vector<int>>{{1, 0, 0, 1}, {0, 1, 1, 0}, {0, 1, 1, 1}, {1, 0, 1, 1}}) << " -> 1" << endl;
    cout << cls.findCircleNum2(vector<vector<int>>{{1, 1, 0}, {1, 1, 0}, {0, 0, 1}}) << " -> 2" << endl;
    cout << cls.findCircleNum2(vector<vector<int>>{{1, 0, 0}, {0, 1, 0}, {0, 0, 1}}) << " -> 3" << endl;
    return 0;
}

File: go/500.600/547.NumberOfProvinces.go
package main

import (
	"fmt"
)

func findCircleNum(isConnected [][]int) int {
	n := len(isConnected)
	visited := make([]int, n)
	ans := 0
	var dfs func(i int)
	dfs = func(i int) {
		for j := 0; j < n; j++ {
			if isConnected[i][j] == 1 && visited[j] == 0 {
				visited[j] = 1
				dfs(j)
			}
		}
	}
	for i := 0; i < n; i++ {
		if visited[i] == 0 {
			dfs(i)
			ans++
		}
	}
	return ans
}

func findCircleNum1(isConnected [][]int) int {
	n := len(isConnected)
	visited := make([]bool, n)
	ans := 0
	for i, v := range visited {
		if !v {
			ans++
			queue := []int{i}
			for len(queue) > 0 {
				c := queue[0]
				queue = queue[1:]
				visited[c] = true
				for to, conn := range isConnected[c] {
					if conn == 1 && !visited[to] {
						queue = append(queue, to)
					}
				}
			}
		}
	}

	return ans
}

func findCircleNum2(isConnected [][]int) int {
	n := len(isConnected)
	parent := make([]int, n)
	for i := range parent {
		parent[i] = i
	}
	var find func(int) int
	find = func(i int) int {
		if parent[i] != i {
			parent[i] = find(parent[i])
		}
		return parent[i]
	}
	union := func(from, to int) {
		parent[find(from)] = find(to)
	}
	for i, row := range isConnected {
		for j := i + 1; j < n; j++ {
			if row[j] == 1 {
				union(i, j)
			}
		}
	}
	ans := 0
	for i, p := range parent {
		if i == p {
			ans++
		}
	}
	return ans
}

func main() {
	fmt.Println(findCircleNum2([][]int{{1, 0, 0, 1}, {0, 1, 1, 0}, {0, 1, 1, 1}, {1, 0, 1, 1}}), 1)
	fmt.Println(findCircleNum2([][]int{{1, 1, 0}, {1, 1, 0}, {0, 0, 1}}), 2)
	fmt.Println(findCircleNum2([][]int{{1, 0, 0}, {0, 1, 0}, {0, 0, 1}}), 3)
}
